src/backend/distributed/commands/trigger.c

```diff
@@ -203,45 +242,480 @@ get_relation_trigger_oid_compat(HeapTuple heapTuple)
 
-	if (!OidIsValid(relationId))
+		CreateTrigStmt *createTriggerStmt = castNode(CreateTrigStmt, parseTree);
+		if (!IsCreateCitusTruncateTriggerStmt(createTriggerStmt))
+		{
+			checkTableType = true;
+			relation = createTriggerStmt->relation;
+		}
+	}
+	else if (IsA(parseTree, RenameStmt))
+	{
+		/* ALTER TRIGGER oldName ON relationName RENAME TO newName */
+
+		RenameStmt *renameTriggerStmt = castNode(RenameStmt, parseTree);
+		if (renameTriggerStmt->renameType == OBJECT_TRIGGER)
+		{
+			checkTableType = true;
+			relation = renameTriggerStmt->relation;
+		}
+	}
+	else if (IsA(parseTree, AlterObjectDependsStmt))
+	{
+		/* ALTER TRIGGER triggerName ON relationName DEPENDS ON extensionName */
+
+		AlterObjectDependsStmt *alterTriggerDependsStmt =
+			castNode(AlterObjectDependsStmt, parseTree);
+		if (alterTriggerDependsStmt->objectType == OBJECT_TRIGGER)
+		{
+			checkTableType = true;
+			relation = alterTriggerDependsStmt->relation;
+		}
+	}
+	else if (IsA(parseTree, DropStmt))
+	{
+		/* DROP TRIGGER triggerName ON relationName */
+
+		DropStmt *dropTriggerStmt = castNode(DropStmt, parseTree);
+		if (dropTriggerStmt->removeType == OBJECT_TRIGGER)
+		{
+			checkTableType = true;
+			relation = GetDropTriggerStmtRelation(dropTriggerStmt);
+		}
+	}
+
+	if (!checkTableType)
 	{
-		/*
-		 * standard_ProcessUtility would already error out if the given table
-		 * does not exist
-		 */
 		return;
 	}
 
+	bool missingOk = false;
+	Oid relationId = RangeVarGetRelid(relation, lockMode, missingOk);
+
 	if (!IsCitusTable(relationId))
 	{
 		return;
 	}
 
-	char *functionName = makeRangeVarFromNameList(createTriggerStmt->funcname)->relname;
+	if (!IsCitusLocalTable(relationId))
+	{
+		const char *relationName = relation->relname;
+		ereport(ERROR, (errmsg("cannot execute CREATE/ALTER/DROP trigger command for "
+							   "relation \"%s\" because it is either a distributed "
+							   "table or a reference table", relationName)));
+	}
+
+	EnsureCoordinator();
+}
+
+
+/*
+ * GetDropTriggerStmtRelation takes a DropStmt for a trigger
+ * object and returns RangeVar for the relation that owns the trigger.
+ */
+static RangeVar *
+GetDropTriggerStmtRelation(DropStmt *dropTriggerStmt)
+{
+	Assert(dropTriggerStmt->removeType == OBJECT_TRIGGER);
+
+	List *targetObjectList = dropTriggerStmt->objects;
+
+	/* drop trigger commands can only drop one at a time */
+	Assert(list_length(targetObjectList) == 1);
+	List *triggerObjectNameList = linitial(targetObjectList);
+
+	List *relationNameList = NIL;
+	SplitLastPointerElement(triggerObjectNameList, &relationNameList, NULL);
+
+	return makeRangeVarFromNameList(relationNameList);
+}
+
+
+/*
+ * PostprocessCreateTriggerStmt is called after a CREATE TRIGGER command has
+ * been executed by standard process utility. This function errors out for
+ * unsupported commands or creates ddl job for supported CREATE TRIGGER commands.
+ */
+List *
+PostprocessCreateTriggerStmt(Node *node, const char *commandString)
```

Comment:
The reason we call this `PostprocessCreateTriggerStmt` seems to be that this is post process based on Postgres, and this is PreProcess if we think it from Citus perspective, so I would prefer `CitusPreProcessCreateTriggerStmt` or something like that for better understandability. What do you think?

Reply:
I think we should prefer the convention in `distribute_object_ops.c`